### DISCUSSION TITLE
[13일차-윤경] 타입과 인터페이스의 차이

### DISCUSSION THREAD

**Discussion by skyhanull** [zzangCS/zzangInterview#43] [프론트 면접 질문]
타입과 인터페이스의 차이에 대해 설명해주세요

**Comment by skyhanull** [comment]
type과 interface의 차이란?
type과 interface 모두 typescript에서 타입을 지정해줄 때 사용합니다
이 두가지는 거의 비슷하게 사용되지만 약간의 차이점이 존재합니다.

확장방법
interface는 타입을 확장할 때 extends를 써서 확장을 합니다.
type는 타입을 확장할 때 &를 써서 확장을 합니다.

선언적확장
interface에서 할 수 있는 대부분의 기능들은 type에서 가능하지만, 한 가지 중요한 차이점은 type은 새로운 속성을 추가하기 위해서 다시 같은 이름으로 선언할 수 없지만, interface는 항상 선언적 확장이 가능하다는 것이다.

computed property name사용유무
interface는 객체로 사용이 가능하지만 computed property name은 사용이 불가능합니다
하지만 type은 computed property name사용이 가능합니다
computed property name은 인덱스 시그니쳐와 다른것입니다!

**Comment by ez1n** [comment]
type과 interface : 객체의 타입의 이름을 지정하는 것

type
- 특수문자 &로 확장 (상속)
- 새로운 속성을 추가하기 위해 다시같은 이름으로 선언할 수 없음
- computed value 사용 가능

interface
- extends로 확장
- 선언적 확장 가능 (같은 이름의 객체를 다시 선언하면 자동으로 확장됨)
- 객체에만 사용 가능 (리터럴 타입으로 확장할 수 없음)
- computed value 사용 불가

외부에 공개되어야 하는 library같은 경우 타입 객체의 확장성을 위해 interface를 권장한다고 함
튜플이나 단순한 원시값의 타입 선언의 경우 type을 사용하는 것을 권장
